### PR TITLE
feat: listening events from kubernetes

### PR DESCRIPTION
**Pull request by kilosonc**
### Description of your changes

1. Add k8s event package to listen for k8s events and write them to the database.
2. Use informer to list all pods in the resourcetree, greatly reducing the pressure on the kube API server.
3. Add a clean job to periodically clean up webhook logs and events.

Fixes https://github.com/horizoncd/horizon/issues/147

I have:

- [x] Read and followed Horizon's [contribution process](https://github.com/horizoncd/horizon/CONTRIBUTING.md).
- [x] Run `make build && make lint` to ensure this PR is ready for review.

### How has this code been tested

Has been tested on dev environment

### Special notes for your reviewer

@iamyeka @bysph

### CODE REVIEW

File: pkg/cd/cd.go [hunk 1]
```diff
@@ -88,18 +90,21 @@ type CD interface {
 }
 
 type cd struct {
-	kubeClientFty  kubeclient.Factory
-	factory        argocd.Factory
-	clusterGitRepo gitrepo.ClusterGitRepo
-	targetRevision string
+	kubeClientFty     kubeclient.Factory
```
Comment (cubxxw) [inline pkg/cd/cd.go:93]: kubeClientFty = kubeClientFactory?

File: pkg/config/k8sevent/k8sevent.go [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+type Reason struct {
+	Reason   string   `yaml:"reason"`
+	Messages []string `yaml:"messages"`
+}
+
+type Rule struct {
+	schema.GroupVersionKind `yaml:",inline"`
```
Comment (cubxxw) [inline pkg/config/k8sevent/k8sevent.go:13]:
```suggestion
	schema.GroupVersionKind `yaml:",inline"`
```
Nested? 🤔

File: pkg/event/dao/dao.go [hunk 3]
```diff
@@ -120,4 +134,40 @@ func (d *dao) GetCursor(ctx context.Context) (*models.EventCursor, error) {
...
+	}
+
+	if result := tx.Where("event_id in (?)", id).Delete(&webhookmodels.WebhookLog{}); result.Error != nil {
+		tx.Rollback()
+		return 0, herrors.NewErrDeleteFailed(herrors.WebhookLogInDB, result.Error.Error())
+	}
+
+	tx.Commit()
+	return result.RowsAffected, nil
+}
```
Comment (cubxxw) [inline pkg/event/dao/dao.go:171]:
```suggestion
func (d *dao) DeleteEvents(ctx context.Context, id ...uint) (int64, error) {
	var events []*models.Event
	tx := d.db.WithContext(ctx).Begin()

	result := tx.Where("id in (?)", id).Delete(&events)
	if result.Error != nil {
		tx.Rollback()
		return 0, herrors.NewErrDeleteFailed(herrors.EventInDB, result.Error.Error())
	}

	if result := tx.Where("event_id in (?)", id).Delete(&webhookmodels.WebhookLog{}); result.Error != nil {
		tx.Rollback()
		return 0, herrors.NewErrDeleteFailed(herrors.WebhookLogInDB, result.Error.Error())
	}

	tx.Commit()
	return result.RowsAffected, nil
}
```
1. uint:
```go
if len(id) == 0 ?
```

2. What about the delete event timeout?

File: pkg/jobs/eventhandler/eventhandler.go [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+	eventhandlercfg "github.com/horizoncd/horizon/pkg/config/eventhandler"
+	eventhandlersvc "github.com/horizoncd/horizon/pkg/eventhandler"
+	"github.com/horizoncd/horizon/pkg/jobs"
+	"github.com/horizoncd/horizon/pkg/param/managerparam"
+)
+
+func New(ctx context.Context, eventHandlerConfig eventhandlercfg.Config,
+	mgrs *managerparam.Manager) (jobs.Job, eventhandlersvc.Service) {
+	// start event handler service to generate webhook log by events
+	eventHandlerService := eventhandlersvc.NewService(ctx, mgrs, eventHandlerConfig)
```
Comment (cubxxw) [inline pkg/jobs/eventhandler/eventhandler.go:15]: log

File: pkg/workload/deployment/deployment.go [hunk 5]
```diff
@@ -26,12 +27,32 @@ import (
...
+		Group:    "apps",
+		Version:  "v1",
+		Resource: "replicasets",
+	}
+	GVRPod = schema.GroupVersionResource{
+		Group:    "",
+		Version:  "v1",
+		Resource: "pods",
+	}
 )
```
Comment (cubxxw) [inline pkg/workload/deployment/deployment.go:53]:
```suggestion
var (
	GVRDeployment = schema.GroupVersionResource{
		Group:    "apps",
		Version:  "v1",
		Resource: "deployments",
	}
	GVRReplicaSet = schema.GroupVersionResource{
		Group:    "apps",
		Version:  "v1",
		Resource: "replicasets",
	}
	GVRPod = schema.GroupVersionResource{
		Group:    "",
		Version:  "v1",
		Resource: "pods",
	}
)
```
```go
var GVRs = map[string]GVR{
    "Deployment": {
        Group:    "apps",
        Version:  "v1",
        Resource: "deployments",
    },
...
```

**Review by cubxxw — COMMENTED** [review]
LGTM

File: core/cmd/cmd.go [hunk 6]
```diff
@@ -705,6 +710,9 @@ func Run(flags *Flags) {
 	// enable pprof
 	runPProfServer(&configs.PProf)
 
+	// init log
+	InitLog(flags)
+
```
[the remark this answers is not in the document]
Comment (kilosonc) [reply core/cmd/cmd.go:715]: gin.New() instead, horizon prefer to do not use predefined middleware.

File: pkg/event/dao/dao.go (hunk 3 above)
Comment (kilosonc) [reply pkg/event/dao/dao.go:171]: we have no timeout when operating the db

File: pkg/jobs/clean/cleaner.go [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+			time.Sleep(5 * time.Second)
+			continue
+		}
+		if len(events) == 0 {
+			return
+		}
+
+		maxID := uint(0)
+		needDeleted = needDeleted[:0]
+	OUTTER:
```
Comment (xuzhu-591) [inline pkg/jobs/clean/cleaner.go:141]: typos

File: pkg/jobs/clean/cleaner.go [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+		for _, event := range events {
+			if event.ID > maxID {
+				maxID = event.ID
+			}
+			for _, rule := range c.EventCleanRules {
+				if rule.EventType != event.EventType {
```
Comment (xuzhu-591) [inline pkg/jobs/clean/cleaner.go:147]: extract a method for need clean

File: pkg/workload/kservice/kservice.go [hunk 9]
```diff
@@ -27,14 +27,64 @@ import (
...
+		Group:    "",
+		Version:  "v1",
+		Resource: "pods",
+	}
+)
+
 func init() {
-	workload.Register(ability)
+	workload.Register(ability, GVRService, GVRPod)
+	// schema.GroupVersionResource{
```
Comment (xuzhu-591) [inline pkg/workload/kservice/kservice.go:53]: why comments?

File: pkg/event/manager/manager.go [hunk 10]
```diff
@@ -30,14 +31,18 @@ import (
...
-	ListEvents(ctx context.Context, offset, limit int) ([]*models.Event, error)
+	CreateEvent(ctx context.Context, event ...*models.Event) ([]*models.Event, error)
+	ListEvents(ctx context.Context, query *q.Query) ([]*models.Event, error)
 	ListEventsByRange(ctx context.Context, start, end uint) ([]*models.Event, error)
 	CreateOrUpdateCursor(ctx context.Context,
 		eventIndex *models.EventCursor) (*models.EventCursor, error)
-	GetCursor(ctx context.Context) (*models.EventCursor, error)
+	GetCursor(ctx context.Context, cursorType models.EventCursorType,
+		RegionIDs ...uint) (*models.EventCursor, error)
+	GetCursors(ctx context.Context, cursorType models.EventCursorType,
```
Comment (xuzhu-591) [inline pkg/event/manager/manager.go:41]: the methed is not used.

File: pkg/regioninformers/regioninformers.go [hunk 11]
```diff
@@ -0,0 +1,486 @@
...
+	mu      sync.RWMutex
+	clients map[uint]*RegionClient
+}
+
+// NewRegionInformers is called when initializing
+func NewRegionInformers(regionMgr manager.Manager, defaultResync time.Duration) *RegionInformers {
+	f := RegionInformers{
+		regionMgr:     regionMgr,
+		clients:       make(map[uint]*RegionClient),
+		handlers:      make([]Resource, 0, 16),
```
[the remark this answers is not in the document]
Comment (kilosonc) [reply pkg/regioninformers/regioninformers.go:67]: try to reduce reallocation.

File: pkg/jobs/k8sevent/k8sevent.go [hunk 12]
```diff
@@ -0,0 +1,408 @@
...
+				return nil, err
+			}
+			horizonEvent.ResourceID = cluster.ID
+			return horizonEvent, nil
+		}
+		ownerReferences := un.GetOwnerReferences()
+		if len(ownerReferences) != 0 {
+			ownerReference := ownerReferences[0]
+			gvk = schema.FromAPIVersionAndKind(ownerReference.APIVersion, ownerReference.Kind)
+			name = ownerReference.Name
```
[the remark this answers is not in the document]
Comment (kilosonc) [reply pkg/jobs/k8sevent/k8sevent.go:309]: this method will find resource's father if there's ownerReferences, so no need to have `return` here

File: pkg/jobs/k8sevent/k8sevent.go [hunk 13]
```diff
@@ -0,0 +1,408 @@
...
+	if gvk == GVKApplication {
+		cluster, err := e.mgr.ClusterMgr.GetByName(context.Background(), name)
+		if err != nil {
+			return nil, err
+		}
+		horizonEvent.ResourceID = cluster.ID
+		return horizonEvent, nil
+	}
+
+	for {
```
[the remark this answers is not in the document]
Comment (kilosonc) [reply pkg/jobs/k8sevent/k8sevent.go:276]: need to find parent resource

**Pull request merged**